### ISSUE TITLE
[TT-3362] Security: X-Forwarded-For handling still insecure, remote IP address can be spoofed

### ISSUE THREAD

**Issue by mjpieters** [TykTechnologies/tyk#3654]
**Branch/Environment/Version**
- Branch/Version: Stable
- Environment: OSS

**Describe the bug**
Tyk trusts the X-Forwarded-For header unconditionally, taking the *first* IP address.

**Reproduction steps**

- Put Tyk behind a reverse proxy, configured to send X-Forwarded-For with `<old-value>, REMOTE_ADDR` if the client request has a `X-Forwarded-For` header. This is a common setup.
- Configure API to only accept requests from 172.17.18.19
- Send a request with `X-Forwarded-For: 172.17.18.19`, the reverse proxy forwards this request to Tyk with the header `X-Forwarded-For: 172.17.18.19, <real IP>`
- Observe that Tyk accepts the request.

**Actual behavior**
Tyk accepts the request because it takes the *first* IP address from the header. This is an issue, as trust can only be extended to the *last* N entries of the list of IPs (with N defaulting to 1 if X-Forwarded-For is to be inspected at all).

**Expected behavior**
Tyk should have a *trusted proxy count* option, taking an IP address *counting from the end of the header IP list*, based on the trusted count. The option should default to 1.

For the above example, the real IP will then be set to `<real IP>`, not `172.17.18.19`, because only 1 proxy layer was expected and trusted.

The trusted proxy count value can then be set based on the number of hops used in the specific setup of the deployed Tyk gateway.

**At the very least** tyk should not, by default, use X-Forwarded-For information to set an IP address, the current behaviour means that any and all IP-based features are broken and can't be relied upon for security.

**Comment by agata-wit** [comment]
@mjpieters thanks for reaching out with that issue. I'll take it up with the Tyk Gateway squad and get back to you when I have any info of when/if this behaviour could be changed.

**Comment by andyo-tyk** [comment]
Hi @mjpieters,
I'm sorry nobody has come back to you on this report for way too long.
We did check this reported behaviour and Tyk is actually working as designed.
I understand the concept you describe and we will consider this as a potential feature enhancement to the gateway, alongside other developments in this area.
Thank you for your support.

**Comment by mjpieters** [comment]
>  We did check this reported behaviour and Tyk is actually working as designed.

I had checked too, by reading the Tyk source code, but that design is a _security issue_, because this lets you bypass any IP-based security.

This includes:

- bypassing the IP whitelist and blacklist middlewares
- spoofing the IP address in all logging operations
- Overloading the Redis cache by sending spoofed IP addresses without keys

and whatever custom middleware you install that relies on the remote IP address being reliable.

**Comment by mjpieters** [comment]
I see that there _was_ previously a bug report for this at https://github.com/TykTechnologies/tyk/issues/624, which was automatically closed as 'stale', but it actually [included the same solution I propose here](https://github.com/TykTechnologies/tyk/issues/624#issuecomment-325682860):

> The standard way should be based on the number of proxies you trust. TrustProxy = 0 would mean that you don't trust any XFF data. TrustProxy = 1 would mean you trust the last proxy in the chain, etc.

It was even on the planning board for a few releases.

**Comment by buger** [comment]
Just to clarify. In general `X-Forwarded-For` concept means that we trust the underlying proxy. So if you are behind some LB, and even try to spoof this header, LB will override it in any case. 
But if you are sending this heade, without having LB which handles it for you, then yes it could become security issue. 

So I wonder if question is about `X-Forwarded-For` header functionality details, or adding an option to disable reading any data from `X-Forwarded-For` at all?

**Comment by mjpieters** [comment]
Yes, the minimal mitigation would be to disable getting the IP address from `X-Forwarded-For`, unless explicitly enabled. This at least prevents IP address spoofing when **not** behind a reverse proxy.

However, please don't outsource trust to the proxy, it is easy enough to get the configuration there wrong. Loads of tutorials tell you to just append the incoming client IP to an existing `X-Forwarded-For` header, and some reverse proxies come with that configuration by default, or make it very simple to generate the header in that manner (e.g. NGINX has `X-Forwarded-For: $proxy_add_x_forwarded_for`).

Please follow the advice [given by MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For#selecting_an_ip_address): either have a configurable trusted proxy count, what I proposed here, and / or have a trusted proxy IP address list.

While we are here, perhaps you could put support for the newer [`Forwarded` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) on your roadmap too, which is the more expressive replacement for `X-Forwarded-For`. NGINX has some documentation on [how to use that header with a shared secret](https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/), which would at least give you a better trust foundation until [HTTP message signing](https://httpwg.org/http-extensions/draft-ietf-httpbis-message-signatures.html) has been standardised.

**Comment by andyo-tyk** [comment]
Sorry for being so slow getting back to you on this.

We really appreciate your detailed assessment and sharing of your expert knowledge on this - I've raised tickets internally to ensure that we address the security concerns you've raised.

Thank you again for your patience and support of Tyk!

**Comment by andyo-tyk** [comment]
Hopefully better late than never, we are adding an option to select the XFF depth via a new Gateway configuration option in the upcoming Tyk 5.11.0 release. Please keep an eye open for the release notes with this release.
